lib/netext/resolver.go

```diff
@@ -0,0 +1,195 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netext
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+)
+
+// MultiResolver returns all IP addresses for the given host.
+type MultiResolver func(host string) ([]net.IP, error)
+
+// Resolver is an interface that returns DNS information about a given host.
+type Resolver interface {
+	LookupIP(host string) (net.IP, error)
+}
+
+type resolver struct {
+	resolve     MultiResolver
+	selectIndex lib.DNSSelect
+	policy      lib.DNSPolicy
+	rrm         *sync.Mutex
+	rand        *rand.Rand
+	roundRobin  map[string]uint8
+}
+
+type cacheRecord struct {
+	ip4        []net.IP
+	ip6        []net.IP
+	lastLookup time.Time
+}
+
+type cacheResolver struct {
+	resolver
+	ttl   time.Duration
+	cm    *sync.Mutex
+	cache map[string]cacheRecord
+}
+
+// NewResolver returns a new DNS resolver. If ttl is not 0, responses
+// will be cached per host for the specified period. The IP returned from
+// LookupIP() will be selected based on the given sel and pol values.
+func NewResolver(
+	actRes MultiResolver, ttl time.Duration, sel lib.DNSSelect, pol lib.DNSPolicy,
+) Resolver {
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	res := resolver{
+		resolve:     actRes,
+		selectIndex: sel,
+		policy:      pol,
+		rrm:         &sync.Mutex{},
+		rand:        r,
+		roundRobin:  make(map[string]uint8),
+	}
+	if ttl == 0 {
+		return &res
+	}
+	return &cacheResolver{
+		resolver: res,
+		ttl:      ttl,
+		cm:       &sync.Mutex{},
+		cache:    make(map[string]cacheRecord),
+	}
+}
+
+// LookupIP returns a single IP resolved for host, selected according to the
+// configured select and policy options.
+func (r *resolver) LookupIP(host string) (net.IP, error) {
+	ips, err := r.resolve(host)
+	if err != nil {
+		return nil, err
+	}
+
+	ip4, ip6 := groupByVersion(ips)
+	var ip net.IP
+	ip, ips = r.applyPolicy(host, ip4, ip6)
+	if ip != nil {
+		return ip, nil
+	}
+
+	return r.selectOne(host, ips), nil
+}
+
+// LookupIP returns a single IP resolved for host, selected according to the
+// configured select and policy options. Results are cached per host and will be
+// refreshed if the last lookup time exceeds the configured TTL (not the TTL
+// returned in the DNS record).
+func (r *cacheResolver) LookupIP(host string) (net.IP, error) {
+	r.cm.Lock()
+
+	var ips, ip4, ip6 []net.IP
+	// TODO: Invalidate? When?
+	if cr, ok := r.cache[host]; ok && time.Now().Before(cr.lastLookup.Add(r.ttl)) {
+		ip4, ip6 = cr.ip4, cr.ip6
+	} else {
+		r.cm.Unlock() // The lookup could take some time, so unlock momentarily.
+		var err error
+		ips, err = r.resolve(host)
+		if err != nil {
+			return nil, err
+		}
+		ip4, ip6 = groupByVersion(ips)
+		r.cm.Lock()
+		r.cache[host] = cacheRecord{ip4: ip4, ip6: ip6, lastLookup: time.Now()}
+	}
+
```

Comment:
As we discussed with @na-- previously the "policy" can be implemented here:
if you have `prefer<version>` if there are IPs from that version *only* they should be saved ... if not only the other ones,
In the case of `only<version>` you only look at .. well that one version
in case of `any` ... you just don't separate them.

This current code might come in handy once we have multiple IPs support as it looks (currently) that each VU will just have 1 IP that will have for it's whole time so you could get 1 VU with ipv4 and another with IPv6 and in the case of `preferIPv4` the second one ... still needs to get IPv6 (otherwise it will just not work ;) ) ... but I would argue that then ... the policy should be changed to an `only<version>` given that the VU in question can't actually make other requests

Reply:
I didn't get the scenario in your second paragraph, but agree that it can be simplified. Will change.

Reply:
after https://github.com/loadimpact/k6/pull/1658 (or anything like it) each VU will have an IP that it will use (unless we completely redesign it ... but having multiple IPs have ... problems). So it is possible that you will have VUs that have an ipv4 and VUs which have ipv6 addresses in the same run. But obviously if you have a VU which will use ipv4 .. it should *not* get an ipv6 address even if `preferIPv6` is configured ... 

But this case IMO will be better fixed in v0.30.0

Reply:
Let me know if e0480405 resolves this.

The changes in #1658 will indeed need to be reconciled with the DNS `policy` option. There's also the case if `onlyIPv6` is specified, then it doesn't make sense for the VU to use an outgoing IPv4 interface. But yeah, let's leave that for a follow-up PR.